torchtune/modules/tokenizers/_utils.py

```diff
@@ -4,58 +4,178 @@
         """
         pass
 
 
-def _split_long_repetitions(s: str, max_consecutive_slice_len: int) -> Iterator[str]:
-    """
-    Split the string `s` so that each substring contains no more than `max_consecutive_slice_len`
-    consecutive whitespaces or consecutive non-whitespaces
+def tokenize_messages_no_special_tokens(
+    tokenizer: ModelTokenizer,
+    messages: List[Message],
+    bos_id: int,
+    eos_id: int,
+    max_seq_len: Optional[int] = None,
+) -> Tuple[List[int], List[bool]]:
+    r"""Tokenize a list of messages one at a time then concatenate them,
+    returning a list of tokens and a list of masks. Does not add any special
+    tokens except for BOS and EOS. This serves as a common starting point for
+    model tokenizers that do not rely heavily on special tokens.
+
+    Example:
+        >>> messages = [
+            Message(role="system", content="system message\n", masked=True),
+            Message(role="user", content="user prompt\n", masked=True),
+            Message(role="assistant", content="assistant response\n"),
+        ]
+        # tokenize_messages encodes messages separately and concats
+        >>> tokenize_messages_no_special_tokens(
+            tokenizer,
+            messages,
+            tokenizer.bos_id,
+            tokenizer.eos_id,
+            max_seq_len
+        )[0]
+        [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+        # Same result as encoding the full string in one go
+        >>> tokenizer.encode(''.join([message.content for message in messages]))
+        [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+    Args:
+        messages (List[Message]): A list of messages, each containing role, content,
+            and masked attributes.
+        max_seq_len (Optional[int]): A max sequence length to truncate tokens to.
+            Default: None
+
+    Returns:
+        Tuple[List[int], List[bool]]: The tokenized messages
     """
-    current_slice_len = 0
-    current_slice_is_space = s[0].isspace() if len(s) > 0 else False
-    slice_start = 0
+    start_of_turn = True
+    end_of_turn = False
+    prev_ends_with_space = False
+    tokenized_messages = []
+    mask = []
+    for message in messages:
+        # If assistant message, this is the end of a turn
+        end_of_turn = message.role == "assistant"
 
-    for i in range(len(s)):
-        is_now_space = s[i].isspace()
+        # Prepend BOS on start of new turns
+        if start_of_turn:
+            tokenized_messages.append(bos_id)
+            mask.append(message.masked)
 
-        if current_slice_is_space ^ is_now_space:
-            current_slice_len = 1
-            current_slice_is_space = is_now_space
+        # We want to trim leading whitespace on the next message when
+        # (a) it is a continuation of the turn (i.e. not the first message)
+        # (b) the vocabulary explicitly encodes whitespace characters, and
```

Comment:
(b) is no longer checked here. Personally I think it's OK to leave this in as we want to be explicit about what we're doing here, but would at least update this comment to be a bit more clear about the fact that this should be checked inside the base tokenizer's encode method